yjit/src/cruby.rs

```diff
@@ -195,6 +249,30 @@ pub struct rb_execution_context_struct {
 /// Pointer to an execution context (EC)
 pub type EcPtr = *const rb_execution_context_struct;
 
+/// Opaque call-data type
+#[repr(C)]
+pub struct rb_call_data {
+    _data: [u8; 0],
+    _marker:
+    core::marker::PhantomData<(*mut u8, core::marker::PhantomPinned)>,
+}
+
+/// Opaque call-info type
+#[repr(C)]
+pub struct rb_callinfo {
+    _data: [u8; 0],
+    _marker:
+    core::marker::PhantomData<(*mut u8, core::marker::PhantomPinned)>,
+}
+
+/// Opaque call-info type
+#[repr(C)]
+pub struct rb_callable_method_entry_t {
+    _data: [u8; 0],
+    _marker:
+    core::marker::PhantomData<(*mut u8, core::marker::PhantomPinned)>,
+}
```

Comment:
Did you have to manually define these? What's the PhantomData for?

Reply:
I had previously been defining my opaque types as just pointer types. I'm following Alan's lead here from ISEQs, though I should probably understand better what's going on. The PhantomData allows adding the PhantomPinned marker, which makes the types immovable (by default, Rust can mov stuff around: "https://doc.rust-lang.org/std/pin/index.html".) I'm trying tell if that's required here or if I could skip it...

As far as whether I had to define them... I'll check what bindgen emits for opaque-but-not-blocklist types. It may be hard to make these opaque types without having bindgen create a bunch of other transitive references, though.

Reply:
Hm. Alan's using a different method to import ruby_value_type. So maybe that's preventing using other Bindgen methods of dealing with enums? I'll try doing it the other way.

Reply:
The psuodo struct part isn't required and we could use a `*cosnt c_void` if we want. The main benefit of definite a psudo struct is that we don't have to cast to and from `c_void` and we avoid type confusion with two different opaque pointers. We could make a macro to make defining these to cut down on noise. It sucks that Rust ergo for this is bad but as a silver lining a language feature is coming to alleviate this : https://github.com/rust-lang/rust/issues/43467

So as far as knowing the details of these structs, we are really only saying that pointer to one is distinct from the other. The phantom data stuff is opting out of traits that are automatically implemented for plain structs to prevent misuse with generic methods in the standard library. I'm mostly echoing what the Nomicon is saying here: https://doc.rust-lang.org/nomicon/ffi.html#representing-opaque-structs

Reply:
I'd argue that the more complex struct definition is a small cost for not being able to accidentally pass the wrong struct to an extern func. We can also wrap something raw (*const c_void or usize) like we currently do with VALUE or CfpPtr. Though my gut reaction would be that I should probably convert CfpPtr to this style of struct instead of the other way.

Reply:
Why convert to these kinds of structs? The newtype wrapper seems just simpler, a lot less code?

Reply:
I'm fine with just wrapping a pointer or integer like CfpPtr. Main difference is that "* const rb_opaque_whatsit as *const u8" vs "OpaqueWhatsitPtr.unwrap() as *const u8". We *do* keep the Ruby struct names with the opaque structs, but as you say, the declaration is larger.

Though I suppose we could split the difference with "rb_opaque_whatsit.unwrap() as *const u8". That reads weirder to me, but I'd get used to it.

The other direction "*const u8 as *const rb_opaque_whatsit" presumably turns into "rb_opaque_whatsit(my_ptr)" or similar. That also reads weirdly to me, but I'd get used to that, too.

I think the amount of code is similar, though -- opaque struct versus having to write unwrap (or use a destructuring assignment each time.)